unblob/handlers/compression/bzip2.py

```diff
@@ -0,0 +1,128 @@
+
+from structlog import get_logger
+
+from ...file_utils import Endian, bits, round_up
+from ...models import StructHandler, ValidChunk
+
+logger = get_logger()
+
+BLOCK_HEADER_HI = 0x00003141
+BLOCK_HEADER_LO = 0x59265359
+
+BLOCK_ENDMARK_HI = 0x00001772
+BLOCK_ENDMARK_LO = 0x45385090
+
+FOOTER_SIZE = 10  # 6 bytes magic + 4 bytes CRC
+
+
+class BZip2Handler(StructHandler):
+
+    NAME = "bzip2"
+
+    YARA_RULE = r"""
+        strings:
+            // magic + version + block_size + compressed_magic
+            $magic = /\x42\x5a\x68[\x31-\x39]\x31\x41\x59\x26\x53\x59/
+        condition:
+            $magic
+    """
+
+    C_DEFINITIONS = r"""
+        struct bzip2_header {
+            char magic[2];              // 'BZ' signature/magic number
+            uint8 version;              // 'h' for Bzip2 ('H'uffman coding), '0' for Bzip1 (deprecated)
+            uint8 hundred_k_blocksize;  // '1'..'9' block-size 100 kB-900 kB (uncompressed)
+            char compressed_magic[8];   // 0x314159265359 (BCD (pi))
+            uint32 crc;                 // checksum for this block
+            uint8 randomised;           // 0=>normal, 1=>randomised (deprecated)
+        };
+    """
+    HEADER_STRUCT = "bzip2_header"
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Optional[ValidChunk]:
+
+        self.parse_header(file, Endian.BIG)
+        end_offset = self.bzip2_recover(file, start_offset)
+
+        return ValidChunk(start_offset=start_offset, end_offset=end_offset)
+
+    def bzip2_recover(self, file: io.BufferedIOBase, start_offset: int):  # noqa: C901
+        """Emulate the behavior of bzip2recover, matching on compressed magic and end of stream
+        magic to identify the end offset of the whole bzip2 chunk.
+        """
+
+        bits_read = 0
+        buff_hi = 0
+        buff_lo = 0
+        curr_block = 0
+        rb_ctr = 0
+        block_start = []
+        block_end = []
+        rb_start = []
+        rb_end = []
+
+        file.seek(start_offset)
+        block_start.append(0)
+
+        for b in bits(file):
+            bits_read += 1
+            if b == 2:
+                if (bits_read >= block_start[curr_block]) and (
+                    bits_read - block_start[curr_block] >= 40
+                ):
+                    block_end[curr_block] = bits_read - 1
+                    if curr_block > 0:
+                        logger.debug(
+                            "bzip2_recover (incomplete block)",
+                            block_id=curr_block,
+                            block_start=block_start[curr_block],
+                            block_end=block_end[curr_block],
+                        )
+                else:
+                    curr_block -= 1
+                break
+
+            buff_hi = (buff_hi << 1) | (buff_lo >> 31)
+            buff_hi = buff_hi & 0xFFFFFFFF
+            buff_lo = (buff_lo << 1) | (b & 1)
+            buff_lo = buff_lo & 0xFFFFFFFF
```

Comment:
This whole lo-hi stuff can be significantly improved as Python's int is arbitrary precision so holding 6 bytes there shouldn't be an issue.
```suggestion
            buff = (buff << 1 | b) & 0xFFFF_FFFF_FFFF 
```
Having a single buffer would reduce clutter here significantly. On the other hand I'd like to look into how this code performs to see if it is even feasible to implement it in Python